### PR TITLE
cake 5: Add native type hints for Routing

### PR DESCRIPTION
**Pull request by othercorey**
@markstory @ADmad I forgot Routing.

### CODE REVIEW

File: src/Routing/RouteBuilder.php [hunk 1]
```diff
@@ -914,18 +912,12 @@ public function plugin(string $name, $options = [], $callback = null)
...
      */
-    public function scope(string $path, $params, $callback = null)
+    public function scope(string $path, callable|array $params, ?callable $callback = null)
     {
-        if (!is_array($params)) {
+        if (is_callable($params)) {
             $callback = $params;
             $params = [];
         }
-        if (!is_callable($callback)) {
```
Comment (markstory) [inline src/Routing/RouteBuilder.php:923]: Don't we still need this as $params could be an array, and $callback could be null.

File: src/Routing/Route/EntityRoute.php [hunk 2]
```diff
@@ -65,10 +65,10 @@ public function match(array $url, array $context = []): ?string
      * Checks that we really deal with an entity object
      *
      * @throws \RuntimeException
-     * @param \ArrayAccess|array $entity Entity value from the URL options
+     * @param mixed $entity Entity value from the URL options
      * @return void
      */
-    protected function _checkEntity($entity): void
+    protected function _checkEntity(mixed $entity): void
```
[the remark this answers is not in the document]
Comment (othercorey) [reply src/Routing/Route/EntityRoute.php:71]: This is testing the type in an array. If we type the check function, then it just throws a TypeError and never runs the actual check.

File: src/Routing/RouteBuilder.php (hunk 1 above)
Comment (othercorey) [reply src/Routing/RouteBuilder.php:923]: that's true.

File: src/Routing/Route/EntityRoute.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (ADmad) [reply src/Routing/Route/EntityRoute.php:71]: Ah, makes sense.

**Pull request merged**